tensorflow/contrib/seq2seq/python/ops/seq2seq.py

```diff
@@ -0,0 +1,111 @@
+# Copyright 2016 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+"""Seq2seq layer operations for use in neural networks.
+"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from tensorflow.python.ops import tensor_array_ops
+from tensorflow.python.ops import array_ops
+from tensorflow.python.ops import variable_scope as vs
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import control_flow_ops
+from tensorflow.python.ops import rnn
+
+from tensorflow.contrib import layers
+
+__all__ = ["rnn_decoder"]
+
+def rnn_decoder(cell, decoder_fn, inputs=None,
+                    sequence_lengths=None,
+                    parallel_iterations=None,
+                    swap_memory=False, time_major=False, scope=None):
+  if inputs is not None:
+    # Testing input dimensions
+    if inputs.get_shape().ndims is not None and inputs.get_shape().ndims < 3:
+      raise ValueError("Inputs must have at least three dimensions")
+    if inputs.get_shape()[-1] is None:
+      raise ValueError("Inputs must not be `None` in the feature (3'rd) "
+                       "dimension")
+    # Setup of RNN (dimensions, sizes, length, initial state, dtype)
+    if not time_major:
+      # [batch, seq, features] -> [seq, batch, features]
+      inputs = array_ops.transpose(inputs, perm=[1, 0, 2])
+
+    dtype = inputs.dtype
+    # Get data input information
+    input_depth = int(inputs.get_shape()[2])
+    max_time = array_ops.shape(inputs)[0]
```

Comment:
try to get this statically first:

max_time = inputs.get_shape()[0]
if max_time is None:
  max_time = array_ops.shape(...)[0]